api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -336,17 +338,24 @@ private void checkAndAddPartitionName(String name) {
       checkAndAddPartitionName(name, null);
     }
 
+    Builder checkConflicts(boolean check) {
+      checkConflicts = check;
+      return this;
+    }
+
     private void checkAndAddPartitionName(String name, Integer sourceColumnId) {
       Types.NestedField schemaField = schema.findField(name);
-      if (sourceColumnId != null) {
-        // for identity transform case we allow  conflicts between partition and schema field name as
-        //   long as they are sourced from the same schema field
-        Preconditions.checkArgument(schemaField == null || schemaField.fieldId() == sourceColumnId,
-            "Cannot create identity partition sourced from different field in schema: %s", name);
-      } else {
-        // for all other transforms we don't allow conflicts between partition name and schema field name
-        Preconditions.checkArgument(schemaField == null,
-            "Cannot create partition from name that exists in schema: %s", name);
+      if (checkConflicts) {
```

Comment:
Looks like there's no other way around this (even using `add` directly) so I think this is the cleanest way to fix it. Thanks, @huaxingao!